reconcile/sendgrid_teammates.py

```diff
@@ -0,0 +1,98 @@
+import logging
+
+import sendgrid
+
+import reconcile.queries as queries
+from reconcile.utils.secret_reader import SecretReader
+
+LOG = logging.getLogger(__name__)
+QONTRACT_INTEGRATION = 'sendgrid_teammates'
+
+
+class Teammate:
+    def __init__(self, email, pending=False):
+        self.email = email
+        self.pending = pending
+
+
+def fetch_desired_state(users):
+    desired_state = {}
+    for user in users:
+        roles = user.get('roles', [])
+        for role in roles:
+            sendgrid_accounts = role.get('sendgrid_accounts') or []
+            for sg_account in sendgrid_accounts:
+                desired_state.setdefault(sg_account['name'], [])
+                t = Teammate(f"{user['org_username']}@redhat.com")
+                desired_state[sg_account['name']].append(t)
+
+    return desired_state
+
+
+def fetch_current_state(sg_client):
+    state = []
+
+    # pending invites
+    invites = sg_client.teammates.pending.get().to_dict['result']
+    for invite in invites:
+        t = Teammate(invite['email'], pending=True)
+        state.append(t)
+
+    # current teammates
+    teammates = sg_client.teammates.get().to_dict['result']
+    for teammate in teammates:
+        if teammate['user_type'] == 'owner':
+            # we want to ignore the root account (owner account)
+            continue
+
+        t = Teammate(teammate['email'], pending=False)
+        state.append(t)
+
+    return state
+
+
+def act(dry_run, sg_client, desired_state, current_state):
+    desired_emails = [e.email for e in desired_state]
+    current_emails = [e.email for e in current_state]
+
+    for user in current_emails:
+        if user.email not in desired_emails:
+            LOG.info(['delete', user.email])
+
+    for user in desired_state:
+        if user.email not in current_emails:
+            # ignore pending users
+            if user.pending:
+                continue
+
+            LOG.info(['invite', user.email])
+
+            if not dry_run:
+                req = {
+                    'email': user.email,
+                    'scopes': [],
+                    'is_admin': True,
+                }
+
+                response = sg_client.teammates.post(request_body=req)
+                if int(response.status_code / 100) != 2:
+                    LOG.error(['error inviting user',
```

Comment:
in addition to logging an error we should eventually exit with an error

Reply:
thanks for catching that! fixed